src/lsm/compaction.zig

```diff
@@ -126,125 +140,101 @@ pub fn CompactionType(
-                .manifest = undefined,
-                .level_b = undefined,
                 .range = undefined,
                 .snapshot = undefined,
                 .drop_tombstones = undefined,
 
+                .status = .idle,
                 .iterator_a = iterator_a,
                 .iterator_b = iterator_b,
 
-                .merge_iterator = undefined, // This can only be initialized on tick 1.
-                .table_builder = table_builder,
+                .merge_done = false,
+                .merge_iterator = null,
 
-                .index = index,
-                .filter = filter,
-                .data = data,
+                .table_builder = table_builder,
+                .index = .{},
+                .filter = .{},
+                .data = .{},
 
+                // Assigned by start()
+                .manifest = undefined,
+                .level_b = undefined,
+                .remove_level_a = null,
                 .update_level_b = update_level_b,
                 .insert_level_b = insert_level_b,
             };
         }
 
-        fn allocate_block(allocator: mem.Allocator) !BlockPtr {
-            const block = try allocator.alignedAlloc(u8, config.sector_size, config.block_size);
-            return block[0..config.block_size];
-        }
-
         pub fn deinit(compaction: *Compaction, allocator: mem.Allocator) void {
-            compaction.iterator_a.deinit(allocator);
-            compaction.iterator_b.deinit(allocator);
-            compaction.table_builder.deinit(allocator);
-            compaction.update_level_b.deinit(allocator);
             compaction.insert_level_b.deinit(allocator);
+            compaction.update_level_b.deinit(allocator);
+
+            compaction.table_builder.deinit(allocator);
 
-            allocator.free(compaction.index.block);
-            allocator.free(compaction.filter.block);
-            allocator.free(compaction.data.block);
+            compaction.iterator_b.deinit(allocator);
+            compaction.iterator_a.deinit(allocator);
         }
 
+        /// table_a is null when level A is 0.
         pub fn start(
             compaction: *Compaction,
             grid: *Grid,
             manifest: *Manifest,
-            // TODO level_a_table: ?TableInfo,
-            level_b: u8,
-            range: Manifest.CompactionRange,
             snapshot: u64,
+            range: Manifest.CompactionRange,
+            table_a: ?*const TableInfo,
+            level_b: u8,
             iterator_a_context: IteratorA.Context,
         ) void {
             assert(compaction.status == .idle);
             assert(compaction.callback == null);
             assert(compaction.io_pending == 0);
-            assert(level_b < config.lsm_levels);
+            assert(!compaction.merge_done and compaction.merge_iterator == null);
+
             assert(range.table_count > 0);
+            assert(range.table_count <= compaction.update_level_b.array.len);
```

Comment:
Doesn't this need a `+ 1` on the right? I think `CompactionRange` includes the level A table.

Reply:
Good catch. What about ` + @boolToInt(table_a != null)`?

Reply:
I think even when `table_a` is null (i.e. we are compacting the immutable table), from the `CompactionRange`'s point of view it is still counted (see: https://github.com/tigerbeetledb/tigerbeetle/blob/dfb67bd95a528bc9a47ba16ccb6111c39c212707/src/lsm/tree.zig#L551-L557).